### ISSUE TITLE
Python/gevent: Segmentation Fault depending on import ordering

### ISSUE THREAD

**Issue by rodaine** [grpc/grpc#17414]
### What version of gRPC and what language are you using?
 
Python, tested on `grpcio` v1.14.0 and v1.16.1
 
### What operating system (Linux, Windows, …) and version?
 
Ubuntu Trusty
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
Python v2.7.6
 
### What did you do?

Attempting to use grpc with gevent support, using the following script (assume a simple Users service definition with a GetUser by some ID):

```py
import gevent.monkey
gevent.monkey.patch_all()

from grpc.experimental.gevent import init_gevent
init_gevent()

import grpc
from pb.users_pb2 import GetUserRequest
from pb.users_pb2_grpc import UsersStub

channel = grpc.insecure_channel('localhost:9001')

import requests # !!!

client = UsersStub(channel)

req = GetUserRequest()
req.id = long(123)
print client.GetUser(req, timeout=.25)
```

Notice the first import of the `requests` package after monkey patching and creation of the channel. 
 
### What did you expect to see?
 
Either a print out of a GetUserResponse or an exception originating from the users service.

### What did you see instead?

For both versions of the `grpcio` package, no network traffic leaves this script (ie, no request actually reaches the users service).

With `grpcio` v1.14.0, a deadline exceeded exception is returned.

With `grpcio` v1.16.1, a segmentation fault crashes the script.
 
### Anything else we should know about your project / environment?

- Removing `import requests` from the script results in expected behavior.
- Moving `import requests` above the creation of the channel results in expected behavior.
- Moving `import requests` after the `GetUser` call (assuming no error there), results in a seg fault.
- Removing the timeout kwarg in v1.14.0 will block forever.
- The following packages have been tested and shown to trigger this behavior:
  - requests: replaces urllib's SSL context with pyopenssl to enable SNI
  - rtree: C library
  - shapely.geometry: C library
  - boto:  vendors six.moves and executes it at import

**Comment by ericgribkoff** [comment]
Thanks for the report. We are tackling a number of non-gevent issues right now, and the team member who wrote our gevent compatible code is no longer on the team, so it will likely be sometime next month before we can follow-up in detail with this bug report - there have been other recent issues reported with our gevent integration that we will also be sorting out.

From initial searching, it wasn't clear to me that requests is compatible with gevent, given the existence of the grequests package. If you have a pointer to documentation about this that would be helpful.

**Comment by rodaine** [comment]
Sorry for the delay in response! Can't say if it's documented somewhere, but we use requests extensively with gevent. In any case, you can reproduce this with any of the 4 packages listed at the end of the issue.

**Comment by grapemix** [comment]
I used gevent + requests a lot too with no problems.

**Comment by arian487** [comment]
Has there been any resolution to this?  We are trying to use gevent + gunicorn + grpc on an appengine flex app and are also getting segfaults.  Its not clear if requests is the problem though

**Comment by lidizheng** [comment]
@arian487 Are you observing the same exception caused by import ordering? If the segfault is caused by other issue, can you open a separate issue? Also, a reproducible case will be very helpful.
